package-res/resources/web/pentaho/visual/base/View.js

```diff
@@ -92,39 +89,105 @@ define([
     },
 
     /**
-     * Orchestrates the rendering of the visualization.
+     * Orchestrates the rendering of the visualization and is meant to be invoked by the container.
      *
-     * This method executes [_render]{@link pentaho.visual.base.View#_render}
-     * asynchronously and is meant to be invoked by the container.
+     * Executes [_update]{@link pentaho.visual.base.View#_update} asynchronously in
+     * the will/did/rejected event loop associated with an update of a visualization,
+     * and also creates the visualization DOM element the first time it successfully updates.
+     *
+     * In order to get the visualization DOM element,
+     * listen to the {@link pentaho.visual.base.events.DidCreate|"did:create"} event.
      *
      * @return {Promise} A promise that is fulfilled when the visualization
      * is completely rendered. If the visualization is in an invalid state, the promise
      * is immediately rejected.
+     *
+     * @fires "will:update"
+     * @fires "did:update"
+     * @fires "rejected:update"
+     * @fires "did:create"
      */
-    render: function() {
-      var me = this;
+    update: function() {
+      var me = this,
+          model = this.model,
+          isFirstRender = !this._element;
+
       return new Promise(function(resolve, reject) {
-        try {
-          var validationErrors = me._validate();
-          if(!validationErrors) {
-            Promise.resolve(me._render()).then(resolve, reject);
-          } else {
-            reject(validationErrors);
-          }
-        } catch(e) {
-          reject(e.message);
+        var willUpdate;
+        if(model._hasListeners(WillUpdate.type)) {
+          willUpdate = new WillUpdate(model);
+          model._emitSafe(willUpdate);
         }
+
+        var result = willUpdate && willUpdate.isCanceled
+          ? ActionResult.reject(willUpdate.cancelReason)
+          : me._doUpdate();
```

Comment:
Updating a visualization is, in general, an asynchronous operation, so `_doUpdate` should return a `Promise`. Generally, the result of `_doUpdate` should be passed to [Promise.resolve(.)](https://developer.mozilla.org/en/docs/Web/JavaScript/Reference/Global_Objects/Promise/resolve).